src/librustc/mir/transform/dataflow.rs

```diff
@@ -99,39 +82,22 @@ pub enum StatementChange<'tcx> {
     Remove,
     /// Replace with another single statement
     Statement(mir::Statement<'tcx>),
-    /// Replace with a list of statements
-    Statements(Vec<mir::Statement<'tcx>>),
 }
 
-impl<'tcx> StatementChange<'tcx> {
-    fn normalise(&mut self) {
-        let old = ::std::mem::replace(self, StatementChange::None);
-        *self = match old {
-            StatementChange::Statements(mut stmts) => {
-                match stmts.len() {
-                    0 => StatementChange::Remove,
-                    1 => StatementChange::Statement(stmts.pop().unwrap()),
-                    _ => StatementChange::Statements(stmts)
-                }
-            }
-            o => o
-        }
-    }
-}
+pub trait Transfer<'tcx> {
+    type Lattice: Lattice;
 
-pub trait Transfer<'tcx, L: Lattice> {
-    type TerminatorOut;
     /// The transfer function which given a statement and a fact produces a fact which is true
     /// after the statement.
-    fn stmt(&mir::Statement<'tcx>, L) -> L;
+    fn stmt(&self, &mir::Statement<'tcx>, Self::Lattice) -> Self::Lattice;
 
     /// The transfer function which given a terminator and a fact produces a fact for each
     /// successor of the terminator.
     ///
     /// Corectness precondtition:
     /// * The list of facts produced should only contain the facts for blocks which are successors
     /// of the terminator being transfered.
-    fn term(&mir::Terminator<'tcx>, L) -> Self::TerminatorOut;
+    fn term(&self, &mir::Terminator<'tcx>, Self::Lattice) -> Vec<Self::Lattice>;
```

Comment:
This change makes the trait not very reusable for backward analysis, where terminators only have a single edge regardless of how many edges the terminator has in the forward direction. That is the primary motivation for TerminatorOut associated type.